# Web Scrapping Project - We will be scraping from Wikipedia

### First we need to load all the necessary libraries:

In [2]:
# Importing the libraries
import pandas as pd #library to work with data frames
import numpy as np #library to work with arrays
from bs4 import BeautifulSoup as bs #library to scrape from website
import requests #library to make requests to website.

### Then we are requesting the http to get the text of the website

In [3]:
source = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text

### Now all we have to do is using BeautifulSoup to load the data and build the pandas dataframe.

In [4]:
soup = bs(source, 'lxml')
soup.prettify()
header=[]
htemp=soup.find("table").tr

#Getting the header into a list
counter=0
for x in htemp.find_all("th"):
    header.append(x.text)
    
#Getting the body into a list with different columns
bod=soup.find("table")
body=[]
for i in bod.find_all("td"):
    body.append(i.text)
    #print(i.text)
    
Postcode = body[0:len(body):3]
Borough = body[1:len(body):3]
Neighbourhood = body[2:len(body):3]
Neighbourhood= [s.replace('\n', '') for s in Neighbourhood] #Dropping the \n

#Creating Pandas dataframe
df=pd.DataFrame({'Borough':Borough,
                 'Postcode':Postcode,
                 'Neighbourhood':Neighbourhood})

#Reorganizing the dataframe's columns
df=df[['Postcode','Borough','Neighbourhood']]

#Creating index list with 'Not assigned' boroughs
emptybor=df.index[df['Borough']=='Not assigned'].tolist()

#Dropping the emptyboroughs
df=df.drop(emptybor,axis=0)

#reseting the indexes of the new dataframe
df=df.reset_index(drop=True)

# Finding the indexes for "Not assigned" neighbourhoods:
nan=df.index[df['Neighbourhood']=='Not assigned'].tolist()

#modifying the "Not assigned" neighbouhoods to Boroughs
df.iloc[nan,2] =df.iloc[nan,1]

# Using groupby we can obtain the list of neighbourhoods that are in same postcode.
df = df.groupby(['Postcode','Borough'])['Neighbourhood'].apply(list)

# Reseting the indexes
df=df.reset_index()

#looking at the first 5 items of final dataframe
df


Postcode           Borough  \
0        M1B       Scarborough   
1        M1C       Scarborough   
2        M1E       Scarborough   
3        M1G       Scarborough   
4        M1H       Scarborough   
5        M1J       Scarborough   
6        M1K       Scarborough   
7        M1L       Scarborough   
8        M1M       Scarborough   
9        M1N       Scarborough   
10       M1P       Scarborough   
11       M1R       Scarborough   
12       M1S       Scarborough   
13       M1T       Scarborough   
14       M1V       Scarborough   
15       M1W       Scarborough   
16       M1X       Scarborough   
17       M2H        North York   
18       M2J        North York   
19       M2K        North York   
20       M2L        North York   
21       M2M        North York   
22       M2N        North York   
23       M2P        North York   
24       M2R        North York   
25       M3A        North York   
26       M3B        North York   
27       M3C        North York   
28       M3H        North York   
29       M3J        North York   
..       ...               ...   
73       M6C              York   
74       M6E              York   
75       M6G  Downtown Toronto   
76       M6H      West Toronto   
77       M6J      West Toronto   
78       M6K      West Toronto   
79       M6L        North York   
80       M6M              York   
81       M6N              York   
82       M6P      West Toronto   
83       M6R      West Toronto   
84       M6S      West Toronto   
85       M7A      Queen's Park   
86       M7R       Mississauga   
87       M7Y      East Toronto   
88       M8V         Etobicoke   
89       M8W         Etobicoke   
90       M8X         Etobicoke   
91       M8Y         Etobicoke   
92       M8Z         Etobicoke   
93       M9A         Etobicoke   
94       M9B         Etobicoke   
95       M9C         Etobicoke   
96       M9L        North York   
97       M9M        North York   
98       M9N              York   
99       M9P         Etobicoke   
100      M9R         Etobicoke   
101      M9V         Etobicoke   
102      M9W         Etobicoke   

                                         Neighbourhood  
0                                     [Rouge, Malvern]  
1             [Highland Creek, Rouge Hill, Port Union]  
2                  [Guildwood, Morningside, West Hill]  
3                                             [Woburn]  
4                                          [Cedarbrae]  
5                                [Scarborough Village]  
6        [East Birchmount Park, Ionview, Kennedy Park]  
7                    [Clairlea, Golden Mile, Oakridge]  
8    [Cliffcrest, Cliffside, Scarborough Village West]  
9                        [Birch Cliff, Cliffside West]  
10   [Dorset Park, Scarborough Town Centre, Wexford...  
11                                 [Maryvale, Wexford]  
12                                         [Agincourt]  
13           [Clarks Corners, Sullivan, Tam O'Shanter]  
14   [Agincourt North, L'Amoreaux East, Milliken, S...  
15                                   [L'Amoreaux West]  
16                                       [Upper Rouge]  
17                                 [Hillcrest Village]  
18                      [Fairview, Henry Farm, Oriole]  
19                                   [Bayview Village]  
20                          [Silver Hills, York Mills]  
21                           [Newtonbrook, Willowdale]  
22                                  [Willowdale South]  
23                                   [York Mills West]  
24                                   [Willowdale West]  
25                                         [Parkwoods]  
26                                   [Don Mills North]  
27                  [Flemingdon Park, Don Mills South]  
28   [Bathurst Manor, Downsview North, Wilson Heights]  
29                   [Northwood Park, York University]  
..                                                 ...  
73                                [Humewood-Cedarvale]  
74            

In [5]:
# Lets see whats the shape of our final dataframe:
df.shape

(103, 3)

The final dataframe contains 103 rows

## Now we are going to find the lattitude and longitude based on our post_codes:

In [15]:
import sys
!{sys.executable} -m pip install geocoder


You are using pip version 18.1, however version 19.2.3 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.


In [27]:
import geocoder # import geocoder

# initializing the variable to None
lat_lng_coords = None

# loop until we get the coordinates
while(lat_lng_coords is None):
    g = geocoder.google('Mountain View, CA', sensor=True)
    lat_lng_coords = g.latlng

latitude = lat_lng_coords[0]
longitude = lat_lng_coords[1]
latitude

KeyboardInterrupt: 

## geocoder did not quite work. Therefore, below I will use the .csv file to add lattitude and longitude data

In [31]:
import os
os.getcwd()


'/Users/John/Downloads/Data-Scraping---Toronto-neighbourhoods-Wikipedia--master'

In [34]:
latlong=pd.read_csv('Geospatial_Coordinates.csv')
lat=latlong['Latitude']
lng=latlong['Longitude']
df['Latitude']=lat
df['Longitude']=lng
df


Postcode           Borough  \
0        M1B       Scarborough   
1        M1C       Scarborough   
2        M1E       Scarborough   
3        M1G       Scarborough   
4        M1H       Scarborough   
5        M1J       Scarborough   
6        M1K       Scarborough   
7        M1L       Scarborough   
8        M1M       Scarborough   
9        M1N       Scarborough   
10       M1P       Scarborough   
11       M1R       Scarborough   
12       M1S       Scarborough   
13       M1T       Scarborough   
14       M1V       Scarborough   
15       M1W       Scarborough   
16       M1X       Scarborough   
17       M2H        North York   
18       M2J        North York   
19       M2K        North York   
20       M2L        North York   
21       M2M        North York   
22       M2N        North York   
23       M2P        North York   
24       M2R        North York   
25       M3A        North York   
26       M3B        North York   
27       M3C        North York   
28       M3H        North York   
29       M3J        North York   
..       ...               ...   
73       M6C              York   
74       M6E              York   
75       M6G  Downtown Toronto   
76       M6H      West Toronto   
77       M6J      West Toronto   
78       M6K      West Toronto   
79       M6L        North York   
80       M6M              York   
81       M6N              York   
82       M6P      West Toronto   
83       M6R      West Toronto   
84       M6S      West Toronto   
85       M7A      Queen's Park   
86       M7R       Mississauga   
87       M7Y      East Toronto   
88       M8V         Etobicoke   
89       M8W         Etobicoke   
90       M8X         Etobicoke   
91       M8Y         Etobicoke   
92       M8Z         Etobicoke   
93       M9A         Etobicoke   
94       M9B         Etobicoke   
95       M9C         Etobicoke   
96       M9L        North York   
97       M9M        North York   
98       M9N              York   
99       M9P         Etobicoke   
100      M9R         Etobicoke   
101      M9V         Etobicoke   
102      M9W         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                     [Rouge, Malvern]  43.806686 -79.194353  
1             [Highland Creek, Rouge Hill, Port Union]  43.784535 -79.160497  
2                  [Guildwood, Morningside, West Hill]  43.763573 -79.188711  
3                                             [Woburn]  43.770992 -79.216917  
4                                          [Cedarbrae]  43.773136 -79.239476  
5                                [Scarborough Village]  43.744734 -79.239476  
6        [East Birchmount Park, Ionview, Kennedy Park]  43.727929 -79.262029  
7                    [Clairlea, Golden Mile, Oakridge]  43.711112 -79.284577  
8    [Cliffcrest, Cliffside, Scarborough Village West]  43.716316 -79.239476  
9                        [Birch Cliff, Cliffside West]  43.692657 -79.264848  
10   [Dorset Park, Scarborough Town Centre, Wexford...  43.757410 -79.273304  
11                                 [Maryvale, Wexford]  43.750072 -79.295849  
12                                         [Agincourt]  43.794200 -79.262029  
13           [Clarks Corners, Sullivan, Tam O'Shanter]  43.781638 -79.304302  
14   [Agincourt North, L'Amoreaux East, Milliken, S...  43.815252 -79.284577  
15                                   [L'Amoreaux West]  43.799525 -79.318389  
16                                       [Upper Rouge]  43.836125 -79.205636  
17                                 [Hillcrest Village]  43.803762 -79.363452  
18                      [Fairview, Henry Farm, Oriole]  43.778517 -79.346556  
19                                   [Bayview Village]  43.786947 -79.385975  
20                          [Silver Hills, York Mills]  43.757490 -79.374714  
21                           [Newtonbrook, Willowdale]  43.789053 -79.408493  
22                                  [Willowdale South]  43.770120 -79.408493  